backend/eventloop/application_event_loop/application_event_runner_test.go

```diff
@@ -1689,6 +1716,121 @@ var _ = Describe("ApplicationEventLoop Test", func() {
 			Expect(resourcesOut).To(BeEmpty())
 		})
 	})
+
+	Context("Test removeFinalizerIfExist function", func() {
+
+		var (
+			ctx        context.Context
+			k8sClient  *sharedutil.ProxyClient
+			gitopsDepl *managedgitopsv1alpha1.GitOpsDeployment
+			informer   sharedutil.ListEventReceiver
+		)
+
+		BeforeEach(func() {
+			ctx = context.Background()
+			scheme, _, _, workspace, err := tests.GenericTestSetup()
+			Expect(err).To(BeNil())
+
+			gitopsDepl = &managedgitopsv1alpha1.GitOpsDeployment{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "my-gitops-depl",
+					Namespace: workspace.Name,
+					UID:       uuid.NewUUID(),
+				},
+				Spec: managedgitopsv1alpha1.GitOpsDeploymentSpec{
+					Source: managedgitopsv1alpha1.ApplicationSource{
+						RepoURL:        "https://github.com/abc-org/abc-repo",
+						Path:           "/abc-path",
+						TargetRevision: "abc-commit"},
+					Type: managedgitopsv1alpha1.GitOpsDeploymentSpecType_Automated,
+					Destination: managedgitopsv1alpha1.ApplicationDestination{
+						Namespace: "abc-namespace",
+					},
+				},
+			}
+
+			innerClient := fake.NewClientBuilder().WithScheme(scheme).WithObjects(workspace, gitopsDepl).Build()
+
+			informer = sharedutil.ListEventReceiver{}
+			k8sClient = &sharedutil.ProxyClient{
+				InnerClient: innerClient,
+				Informer:    &informer,
+			}
+		})
+
+		It("should remove the finalizer when it is found", func() {
+			testFinalizer := "managed.gitops.test/test"
+			gitopsDepl.Finalizers = append(gitopsDepl.Finalizers, testFinalizer)
+			err := k8sClient.Update(ctx, gitopsDepl)
+			Expect(err).To(BeNil())
+
+			err = removeFinalizerIfExist(ctx, k8sClient, gitopsDepl, testFinalizer)
+			Expect(err).To(BeNil())
+
+			gitopsDeplUpdated := false
+			for _, event := range informer.Events {
+				if event.Action == sharedutil.Update && event.ObjectTypeOf() == "GitOpsDeployment" {
+					gitopsDeplUpdated = true
+				}
+			}
+			Expect(gitopsDeplUpdated).To(BeTrue())
+		})
+
+		It("should not update if the finalizer is not found", func() {
+			err := removeFinalizerIfExist(ctx, k8sClient, gitopsDepl, deletionFinalizer)
+			Expect(err).To(BeNil())
+
+			gitopsDeplUpdated := false
+			for _, event := range informer.Events {
+				if event.Action == sharedutil.Update && event.ObjectTypeOf() == "GitOpsDeployment" {
+					gitopsDeplUpdated = true
+				}
+			}
+			Expect(gitopsDeplUpdated).To(BeFalse())
+		})
+
+		It("should not update if the GitOpsDeployment is already deleted", func() {
+			err := k8sClient.Delete(ctx, gitopsDepl)
+			Expect(err).To(BeNil())
+
+			err = removeFinalizerIfExist(ctx, k8sClient, gitopsDepl, deletionFinalizer)
```

Comment:
Even though it's a little strange, the test should probably set the finalizer on the in-memory `gitopsDepl` before calling `removeFinalizerIfExists` (and not call `k8sClient.Update()`).  This will ensure that the reason it doesn't get updated is because it's been deleted, and not because it doesn't have the finalizer set.